### PR TITLE
Implement cluster file sharing mechanism

### PR DESCRIPTION
**Pull request by limetreeestate**
Signed-off-by: Suchitha Dehigaspitiya <suchithald@gmail.com>

### CODE REVIEW

File: src/server/JasmineGraphInstanceService.cpp [hunk 1]
```diff
@@ -2209,7 +2209,164 @@ void *instanceservicesession(void *dummyPt) {
...
+            int noClusters = stoi(data);
+            graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+            instance_logger.log("Received Graph ID: " + graphID, "info");
+
+            //Get cluster filenames and send to coordinator
+            int count = 0;
+            int orgCount = orgPortList.size();
+            std::thread *workerThreads = new std::thread[orgCount];
+            for (int i = 0; i < orgCount; i++) {
+                workerThreads[count++] = std::thread(JasmineGraphInstanceService::collectBucketstoCoordinator,
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:2260]: Change name from `collectBucketstoCoordinator` to `collectBucketsToCoordinator`.

File: src/server/JasmineGraphInstanceService.cpp [hunk 2]
```diff
@@ -2209,7 +2209,164 @@ void *instanceservicesession(void *dummyPt) {
...
+            string graphID = (data);
+            graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+            instance_logger.log("Received Graph ID: " + graphID, "info");
+
+            //Read cluster count from socket
+            bzero(data, INSTANCE_DATA_LENGTH);
+            read(connFd, data, INSTANCE_DATA_LENGTH);
+            int noClusters = stoi(data);
+            graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+            instance_logger.log("Received Graph ID: " + graphID, "info");
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:2253]: This have to be received cluster count

File: src/server/JasmineGraphInstanceService.cpp [hunk 3]
```diff
@@ -2209,7 +2209,164 @@ void *instanceservicesession(void *dummyPt) {
...
+                                                     sessionargs, orgHostList[i], stoi(orgPortList[i]),
+                                                     stoi(orgPortList[i]), graphID, noClusters);
+                sleep(1);
+            }
+
+            for (int threadCount = 0; threadCount < count; threadCount++) {
+                if(workerThreads[threadCount].joinable()) {
+                    workerThreads[threadCount].join();
+                }
+                std::cout << "Thread [A]: " << threadCount << " joined" << std::endl;
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:2270]: Lets remove the std::cout

File: src/server/JasmineGraphInstanceService.cpp [hunk 4]
```diff
@@ -2209,7 +2209,164 @@ void *instanceservicesession(void *dummyPt) {
...
+            string graphID = (data);
+            graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+            instance_logger.log("Received Graph ID: " + graphID, "info");
+
+            //Read cluster count from socket
+            bzero(data, INSTANCE_DATA_LENGTH);
+            read(connFd, data, INSTANCE_DATA_LENGTH);
+            int noClusters = stoi(data);
+            graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+            instance_logger.log("Received Graph ID: " + graphID, "info");
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:2311]: Change to cluster count

File: src/server/JasmineGraphInstanceService.cpp [hunk 5]
```diff
@@ -2209,7 +2209,164 @@ void *instanceservicesession(void *dummyPt) {
...
+
+            for (int i = 0; i < noClusters; i++) {
+                string filename = "AttrCluster_" + graphID + to_string(i) + ".txt";
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:2314]: Change `AttrCluster_` to `attrcluster_`

File: src/server/JasmineGraphInstanceService.cpp [hunk 6]
```diff
@@ -2209,7 +2209,164 @@ void *instanceservicesession(void *dummyPt) {
...
+                string filepath =
+                        utils.getJasmineGraphProperty("org.jasminegraph.server.instance.entityresolutionfolder") + filename;
+                JasmineGraphInstance::sendFileThroughService(coordinatorHost, stoi(coordinatorPort), filename, filepath);
+            }
+
+        }
+        else if (line.compare(JasmineGraphInstanceProtocol::SHARE_CLUSTER_FILES) == 0) {
+            instance_logger.log("Received : " + JasmineGraphInstanceProtocol::SHARE_BUCKETS_TO_ORG, "info");
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:2322]: `SHARE_BUCKETS_TO_ORG` shouldn't this be share cluster files?

File: src/server/JasmineGraphInstanceService.cpp [hunk 7]
```diff
@@ -2209,7 +2209,164 @@ void *instanceservicesession(void *dummyPt) {
...
+            string graphID = (data);
+            graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+            instance_logger.log("Received Graph ID: " + graphID, "info");
+
+            //Read cluster count from socket
+            bzero(data, INSTANCE_DATA_LENGTH);
+            read(connFd, data, INSTANCE_DATA_LENGTH);
+            int noClusters = stoi(data);
+            graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+            instance_logger.log("Received Graph ID: " + graphID, "info");
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:2359]: Change to cluster count

File: src/server/JasmineGraphInstanceService.cpp (hunk 5 above)
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:2314]: This should be buckets not attribute clusters

File: src/server/JasmineGraphInstanceService.cpp [hunk 8]
```diff
@@ -2209,7 +2209,164 @@ void *instanceservicesession(void *dummyPt) {
...
+
+            for (int i = 0; i < noClusters; i++) {
+                string filename = "AttrCluster_" + graphID + to_string(i) + ".txt";
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:2362]: Change `AttrCluster_` to `attrcluster_`

File: src/server/JasmineGraphInstanceService.cpp [hunk 9]
```diff
@@ -2209,7 +2209,164 @@ void *instanceservicesession(void *dummyPt) {
...
+            instance_logger.log("Received Graph ID: " + graphID, "info");
+
+            //Read cluster count from socket
+            bzero(data, INSTANCE_DATA_LENGTH);
+            read(connFd, data, INSTANCE_DATA_LENGTH);
+            int noClusters = stoi(data);
+            graphID = utils.trim_copy(graphID, " \f\n\r\t\v");
+            instance_logger.log("Received Graph ID: " + graphID, "info");
+
+            for (int i = 0; i < noClusters; i++) {
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:2313]: This `noClusters` should be number of buckets?

File: src/server/JasmineGraphInstanceService.cpp [hunk 10]
```diff
@@ -3318,6 +3475,106 @@ int collectBloomFilters(instanceservicesessionargs *sessionargs,
     return 0;
 }
 
+int JasmineGraphInstanceService::collectBucketstoCoordinator(instanceservicesessionargs *sessionargs,
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:3478]: Change `collectBucketstoCoordinator` to `collectBucketsToCoordinator`

File: src/server/JasmineGraphInstanceService.cpp [hunk 11]
```diff
@@ -3318,6 +3475,106 @@ int collectBloomFilters(instanceservicesessionargs *sessionargs,
...
+
+        bzero(data, INSTANCE_DATA_LENGTH);
+        read(sockfd, data, INSTANCE_DATA_LENGTH);
+        response = (data);
+        response = utils.trim_copy(response, " \f\n\r\t\v");
+
+        if (response.compare(JasmineGraphInstanceProtocol::OK) == 0) {
+            instance_logger.log("Received : " + JasmineGraphInstanceProtocol::OK, "info");
+
+
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:3552]: Delete additional new line

File: src/server/JasmineGraphInstanceService.cpp [hunk 12]
```diff
@@ -3318,6 +3475,106 @@ int collectBloomFilters(instanceservicesessionargs *sessionargs,
...
+            write(sockfd, to_string(server_data_port).c_str(), to_string(server_data_port).size());
+            instance_logger.log("Sent : " + server_data_port, "info");
+
+            //Write graph ID to socket
+            write(sockfd, graphID.c_str(), (graphID).size());
+            instance_logger.log("Sent : Graph ID " + graphID, "info");
+
+            //Write no of clusters to socket
+            write(sockfd, to_string(clusterCount).c_str(), to_string(clusterCount).size());
+            instance_logger.log("Sent : Graph ID " + to_string(clusterCount), "info");
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.cpp:3572]: Change `Sent : Graph ID` to `Sent : Number of clusters `

File: src/server/JasmineGraphInstanceService.h [hunk 13]
```diff
@@ -127,6 +127,10 @@ class JasmineGraphInstanceService {
 
     void clusterFilters(string trainData, int noClusters);
 
+    static int collectBucketstoCoordinator(instanceservicesessionargs *sessionargs,
```
Comment (miyurud) [inline src/server/JasmineGraphInstanceService.h:130]: Change to `collectBucketsToCoordinator`

File: src/server/JasmineGraphServer.cpp [hunk 14]
```diff
@@ -2557,27 +2557,157 @@ int JasmineGraphServer::initiateEntityResolution(vector<pair<string, string>> ho
...
+            response = utils.trim_copy(response, " \f\n\r\t\v");
+            if (response.compare(JasmineGraphInstanceProtocol::SHARE_CLUSTER_FILES) == 0) {
+                loop = true;
+                int clusterCount = 4;
+                //Read coordinator port from socket
+                bzero(data, INSTANCE_DATA_LENGTH);
+                read(sockfd, data, INSTANCE_DATA_LENGTH);
+                string clusterShareHost = (data);
+
+                //Read coordinator data port from socket
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2581]: Remove data word from the comment

File: src/server/JasmineGraphServer.cpp [hunk 15]
```diff
@@ -2557,27 +2557,157 @@ int JasmineGraphServer::initiateEntityResolution(vector<pair<string, string>> ho
...
+                    result_wr = write(sockfd, graphID.c_str(), graphID.size());
+
+                    if(result_wr < 0) {
+                        server_logger.log("Error writing to socket", "error");
+                    }
+
+                    server_logger.log("Sent : Graph ID " + graphID, "info");
+
+                    JasmineGraphServer::collectBucketsToMaster(host, port, designatedWorkerHost,
+                                                               designatedWorkerPort, designatedWorkerPort, graphID, 4,
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2692]: Change `designatedWorkerPort` to `designatedWorkerDataPort`

File: src/server/JasmineGraphServer.cpp [hunk 16]
```diff
@@ -2557,27 +2557,157 @@ int JasmineGraphServer::initiateEntityResolution(vector<pair<string, string>> ho
...
+        //TODO: Wait for coordinator to give instruction to share cluster files
+        loop = false;
+        while(!loop) {
+            bzero(data, 301);
+            read(sockfd, data, 300);
+            response = (data);
+            response = utils.trim_copy(response, " \f\n\r\t\v");
+            if (response.compare(JasmineGraphInstanceProtocol::SHARE_CLUSTER_FILES) == 0) {
+                loop = true;
+                int clusterCount = 4;
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2575]: Shall we get the clusterCount from Frontend?

File: src/server/JasmineGraphServer.cpp [hunk 17]
```diff
@@ -2557,27 +2557,157 @@ int JasmineGraphServer::initiateEntityResolution(vector<pair<string, string>> ho
...
         //TODO: Wait for other orgs to complete comparision and send their results
         //TODO: Combine the recieved results
         //TODO: Send combined results back to each organziation
-    }
-    else {
-        bool loop = false;
-        //TODO: Wait for coordinator to give instruction to share cluster files
+        loop = false;
+
         while (!loop) {
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:2707]: Lets remove this infinite while loop

File: src/server/JasmineGraphServer.cpp [hunk 18]
```diff
@@ -3296,6 +3426,114 @@ int JasmineGraphServer::bucketLocalClusters(std::string host, int port, std::str
...
+
+        bzero(data, INSTANCE_DATA_LENGTH);
+        read(sockfd, data, INSTANCE_DATA_LENGTH);
+        response = (data);
+        response = utils.trim_copy(response, " \f\n\r\t\v");
+
+        if (response.compare(JasmineGraphInstanceProtocol::OK) == 0) {
+            instance_logger.log("Received : " + JasmineGraphInstanceProtocol::OK, "info");
+
+
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:3501]: Remove additional new line

File: src/server/JasmineGraphServer.cpp [hunk 19]
```diff
@@ -3296,6 +3426,114 @@ int JasmineGraphServer::bucketLocalClusters(std::string host, int port, std::str
...
+            //Write org hosts to socket
+            string orgHostListStr = "";
+            for (auto p: orgHostList) {
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:3504]: Change `p` to `host`

File: src/server/JasmineGraphServer.cpp [hunk 20]
```diff
@@ -3296,6 +3426,114 @@ int JasmineGraphServer::bucketLocalClusters(std::string host, int port, std::str
...
+                orgHostListStr += p + "|";
+            }
+            orgHostListStr = orgHostListStr.substr(orgHostListStr.size() - 2);
+            write(sockfd, orgHostListStr.c_str(), orgHostListStr.size());
+            instance_logger.log("Sent : " + orgHostListStr, "info");
+
+            //Write org ports to socket
+            string orgPortListStr = "";
+            for (auto p: orgPortList) {
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:3513]: Change `p` to `port`

File: src/server/JasmineGraphServer.cpp [hunk 21]
```diff
@@ -3296,6 +3426,114 @@ int JasmineGraphServer::bucketLocalClusters(std::string host, int port, std::str
...
+            write(sockfd, to_string(server_data_port).c_str(), to_string(server_data_port).size());
+            instance_logger.log("Sent : " + server_data_port, "info");
+
+            //Write graph ID to socket
+            write(sockfd, graphID.c_str(), (graphID).size());
+            instance_logger.log("Sent : Graph ID " + graphID, "info");
+
+            //Write no of clusters to socket
+            write(sockfd, to_string(noClusters).c_str(), to_string(noClusters).size());
+            instance_logger.log("Sent : Graph ID " + to_string(noClusters), "info");
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:3531]: Change `Graph ID` to `no of clusters`

File: src/server/JasmineGraphServer.cpp [hunk 22]
```diff
@@ -3761,3 +3999,104 @@ void JasmineGraphServer::inDegreeDistribution(std::string graphID) {
...
+            write(sockfd, to_string(orgPort).c_str(), to_string(orgPort).size());
+            instance_logger.log("Sent : " + orgPort, "info");
+
+            //Write graph ID to socket
+            write(sockfd, graphID.c_str(), (graphID).size());
+            instance_logger.log("Sent : Graph ID " + graphID, "info");
+
+            //Write no of clusters to socket
+            write(sockfd, to_string(count).c_str(), to_string(count).size());
+            instance_logger.log("Sent : Graph ID " + to_string(count), "info");
```
Comment (miyurud) [inline src/server/JasmineGraphServer.cpp:4098]: Change `Graph ID ` to `No of clusters`

**Review by miyurud — CHANGES_REQUESTED** [review]
Please address the comments I put in the PR

File: src/server/JasmineGraphInstanceService.cpp (hunk 1 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:2260]: Name change done

File: src/server/JasmineGraphInstanceService.cpp (hunk 2 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:2253]: Code changed to log as cluster count

File: src/server/JasmineGraphInstanceService.cpp (hunk 3 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:2270]: removed

File: src/server/JasmineGraphInstanceService.cpp (hunk 4 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:2311]: Code changed to log as cluster count

File: src/server/JasmineGraphInstanceService.cpp (hunk 6 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:2322]: Changed to `SHARE_CLUSTER_FILES`

File: src/server/JasmineGraphInstanceService.cpp (hunk 7 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:2359]: Code changed to log as cluster count

File: src/server/JasmineGraphInstanceService.cpp (hunk 5 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:2314]: changed filename to buckets

File: src/server/JasmineGraphInstanceService.cpp (hunk 9 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:2313]: The count should be the worker count, variable name changed to reflect that

File: src/server/JasmineGraphInstanceService.cpp (hunk 10 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:3478]: Name change done

File: src/server/JasmineGraphInstanceService.cpp (hunk 11 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:3552]: Deleted

File: src/server/JasmineGraphInstanceService.cpp (hunk 12 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.cpp:3572]: Code changed to log as cluster count

File: src/server/JasmineGraphInstanceService.h (hunk 13 above)
Comment (limetreeestate) [reply src/server/JasmineGraphInstanceService.h:130]: Name change done

File: src/server/JasmineGraphServer.cpp (hunk 14 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:2581]: Removed

File: src/server/JasmineGraphServer.cpp (hunk 17 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:2707]: Removed

File: src/server/JasmineGraphServer.cpp (hunk 18 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:3501]: Removed

File: src/server/JasmineGraphServer.cpp (hunk 19 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:3504]: Renamed to hostIter to not conflict with other variable name

File: src/server/JasmineGraphServer.cpp (hunk 20 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:3513]: Renamed to portIter to not conflict with other variable names

File: src/server/JasmineGraphServer.cpp (hunk 21 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:3531]: Renamed

File: src/server/JasmineGraphServer.cpp (hunk 22 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:4098]: Renamed

**Comment by miyurud** [comment]
@limetreeestate Please push the commit pertaining to addressing the PR comments

File: src/server/JasmineGraphInstanceService.cpp (hunk 1 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 2 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 3 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 4 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 6 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 7 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 12 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp:3567]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 11 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp:3547]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 8 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 9 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 10 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.cpp (hunk 5 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.cpp]: Ok

File: src/server/JasmineGraphInstanceService.h (hunk 13 above)
Comment (miyurud) [reply src/server/JasmineGraphInstanceService.h]: Ok

File: src/server/JasmineGraphServer.cpp (hunk 14 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Ok

File: src/server/JasmineGraphServer.cpp (hunk 17 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Ok

File: src/server/JasmineGraphServer.cpp (hunk 18 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Ok

File: src/server/JasmineGraphServer.cpp (hunk 20 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Ok

File: src/server/JasmineGraphServer.cpp (hunk 21 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Ok

File: src/server/JasmineGraphServer.cpp (hunk 22 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Ok

File: src/server/JasmineGraphServer.cpp (hunk 19 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Ok

File: src/server/JasmineGraphServer.cpp (hunk 15 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:2693]: Change added

File: src/server/JasmineGraphServer.cpp (hunk 16 above)
Comment (limetreeestate) [reply src/server/JasmineGraphServer.cpp:2576]: Cluster count changed to be based on worker count as discussed

Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Ok

File: src/server/JasmineGraphServer.cpp (hunk 15 above)
Comment (miyurud) [reply src/server/JasmineGraphServer.cpp]: Ok

**Pull request merged**